File: Solutions/cpp/taskmanagementsystem/TaskManager.cpp
#include "TaskManager.h"
#include <iostream>
#include <algorithm>

// 构造函数
TaskManager::TaskManager() : taskIdCounter(1), userIdCounter(1) {}

// 析构函数
TaskManager::~TaskManager() {
    for (auto t : tasks) delete t;
    for (auto u : users) delete u;
}

// 用户注册
User* TaskManager::registerUser(const std::string& username, const std::string& email) {
    std::string id = generateUserId();
    User* user = new User(id, username, email);
    users.push_back(user);
    return user;
}

// 移除用户
void TaskManager::removeUser(const std::string& userId) {
    auto it = std::find_if(users.begin(), users.end(),
        [&](User* u) { return u->getUserId() == userId; });
    if (it != users.end()) {
        // 解除任务分配
        for (const std::string& tid : (*it)->getAssignedTasks()) {
            Task* t = findTask(tid);
            if (t) t->setAssignedTo("");
        }
        delete *it;
        users.erase(it);
    }
}

// 创建任务
Task* TaskManager::createTask(const std::string& title, const std::string& description, TaskPriority priority) {
    std::string id = generateTaskId();
    Task* task = new Task(id, title, description, TaskStatus::TODO, priority, 0, "");
    tasks.push_back(task);
    return task;
}

// 移除任务
void TaskManager::removeTask(const std::string& taskId) {
    auto it = std::find_if(tasks.begin(), tasks.end(),
        [&](Task* t) { return t->getTaskId() == taskId; });
    if (it != tasks.end()) {
        // 解除用户分配
        std::string assignee = (*it)->getAssignedTo();
        if (!assignee.empty()) {
            User* u = findUser(assignee);
            if (u) u->removeAssignedTask(taskId);
        }
        delete *it;
        tasks.erase(it);
    }
}

// 分配任务
bool TaskManager::assignTask(const std::string& taskId, const std::string& userId) {
    Task* task = findTask(taskId);
    User* user = findUser(userId);
    if (!task || !user || !user->isActive()) return false;
    // 解除原分配
    std::string prevAssignee = task->getAssignedTo();
    if (!prevAssignee.empty() && prevAssignee != userId) {
        User* prevUser = findUser(prevAssignee);
        if (prevUser) prevUser->removeAssignedTask(taskId);
    }
    task->setAssignedTo(userId);
    user->addAssignedTask(taskId);
    return true;
}

// 更新任务状态
bool TaskManager::updateTaskStatus(const std::string& taskId, TaskStatus status) {
    Task* task = findTask(taskId);
    if (!task) return false;
    // 检查依赖
    if (status == TaskStatus::IN_PROGRESS || status == TaskStatus::COMPLETED) {
        if (!checkDependenciesMet(task)) return false;
    }
    task->setStatus(status);
    return true;
}

// 添加任务依赖
bool TaskManager::addTaskDependency(const std::string& taskId, const std::string& dependencyId) {
    Task* task = findTask(taskId);
    Task* dep = findTask(dependencyId);
    if (!task || !dep || taskId == dependencyId) return false;
    task->addDependency(dependencyId);
    return true;
}

// 添加任务评论
bool TaskManager::addTaskComment(const std::string& taskId, const std::string& comment) {
    Task* task = findTask(taskId);
    if (!task) return false;
    task->addComment(comment);
    return true;
}

// 显示用户任务
void TaskManager::displayUserTasks(const std::string& userId) const {
    const User* user = findUser(userId);
    if (!user) {
        std::cout << "User not found.\n";
        return;
    }
    std::cout << "Tasks for user " << user->getName() << ":\n";
    for (const std::string& tid : user->getAssignedTasks()) {
        const Task* t = findTask(tid);
        if (t) t->displayTaskInfo();
    }
}

// 显示所有任务
void TaskManager::displayAllTasks() const {
    std::cout << "All Tasks:\n";
    for (const Task* t : tasks) {
        t->displayTaskInfo();
    }
}

// 显示所有用户
void TaskManager::displayAllUsers() const {
    std::cout << "All Users:\n";
    for (const User* u : users) {
        u->displayUserInfo();
    }
}

// 按状态获取任务
std::vector<Task*> TaskManager::getTasksByStatus(TaskStatus status) const {
    std::vector<Task*> result;
    for (Task* t : tasks) {
        if (t->getStatus() == status) result.push_back(t);
    }
    return result;
}

// 按优先级获取任务
std::vector<Task*> TaskManager::getTasksByPriority(TaskPriority priority) const {
    std::vector<Task*> result;
    for (Task* t : tasks) {
        if (t->getPriority() == priority) result.push_back(t);
    }
    return result;
}

// 私有：查找用户
User* TaskManager::findUser(const std::string& userId) const {
    auto it = std::find_if(users.begin(), users.end(),
        [&](User* u) { return u->getUserId() == userId; });
    return it != users.end() ? *it : nullptr;
}

// 私有：查找任务
Task* TaskManager::findTask(const std::string& taskId) const {
    auto it = std::find_if(tasks.begin(), tasks.end(),
        [&](Task* t) { return t->getTaskId() == taskId; });
    return it != tasks.end() ? *it : nullptr;
}

// 私有：检查依赖是否满足
bool TaskManager::checkDependenciesMet(const Task* task) const {
    for (const std::string& depId : task->getDependencies()) {
        const Task* dep = findTask(depId);
        if (!dep || dep->getStatus() != TaskStatus::COMPLETED) return false;
    }
    return true;
}

// 私有：生成任务ID
std::string TaskManager::generateTaskId() {
    return "T" + std::to_string(taskIdCounter++);
}

// 私有：生成用户ID
std::string TaskManager::generateUserId() {
    return "U" + std::to_string(userIdCounter++);
}

File: Solutions/golang/taskmanagementsystem/task_manager.go
package taskmanagementsystem

import (
	"sync"
	"time"
)

// TaskManager is the core of the task management system
// It follows the Singleton pattern to ensure a single instance
type TaskManager struct {
	tasks       map[string]*Task
	users       map[string]*User
	taskHistory map[string][]*Task // userId -> completed tasks
	mu          sync.RWMutex
}

var (
	instance *TaskManager
	once     sync.Once
)

// GetInstance returns the singleton instance of TaskManager
func GetInstance() *TaskManager {
	once.Do(func() {
		instance = &TaskManager{
			tasks:       make(map[string]*Task),
			users:       make(map[string]*User),
			taskHistory: make(map[string][]*Task),
		}
	})
	return instance
}

// CreateTask creates a new task
func (tm *TaskManager) CreateTask(task *Task) {
	tm.mu.Lock()
	defer tm.mu.Unlock()
	tm.tasks[task.GetId()] = task
}

// UpdateTask updates an existing task
func (tm *TaskManager) UpdateTask(task *Task) bool {
	tm.mu.Lock()
	defer tm.mu.Unlock()
	if _, exists := tm.tasks[task.GetId()]; exists {
		tm.tasks[task.GetId()] = task
		return true
	}
	return false
}

// DeleteTask deletes a task by ID
func (tm *TaskManager) DeleteTask(taskId string) bool {
	tm.mu.Lock()
	defer tm.mu.Unlock()
	if task, exists := tm.tasks[taskId]; exists {
		// Remove task from assigned user
		if task.GetAssignedTo() != nil {
			task.GetAssignedTo().RemoveAssignedTask(task)
		}
		delete(tm.tasks, taskId)
		return true
	}
	return false
}

// GetTask retrieves a task by ID
func (tm *TaskManager) GetTask(taskId string) *Task {
	tm.mu.RLock()
	defer tm.mu.RUnlock()
	return tm.tasks[taskId]
}

// SearchTasks searches tasks by keyword in title or description
func (tm *TaskManager) SearchTasks(keyword string) []*Task {
	tm.mu.RLock()
	defer tm.mu.RUnlock()

	var results []*Task
	for _, task := range tm.tasks {
		if contains(task.GetTitle(), keyword) || contains(task.GetDescription(), keyword) {
			results = append(results, task)
		}
	}
	return results
}

// FilterTasksByStatus filters tasks by status
func (tm *TaskManager) FilterTasksByStatus(status TaskStatus) []*Task {
	tm.mu.RLock()
	defer tm.mu.RUnlock()

	var results []*Task
	for _, task := range tm.tasks {
		if task.GetStatus() == status {
			results = append(results, task)
		}
	}
	return results
}

// FilterTasksByPriority filters tasks by priority
func (tm *TaskManager) FilterTasksByPriority(priority TaskPriority) []*Task {
	tm.mu.RLock()
	defer tm.mu.RUnlock()

	var results []*Task
	for _, task := range tm.tasks {
		if task.GetPriority() == priority {
			results = append(results, task)
		}
	}
	return results
}

// FilterTasksByAssignedUser filters tasks by assigned user
func (tm *TaskManager) FilterTasksByAssignedUser(userId string) []*Task {
	tm.mu.RLock()
	defer tm.mu.RUnlock()

	var results []*Task
	for _, task := range tm.tasks {
		if task.GetAssignedTo() != nil && task.GetAssignedTo().GetId() == userId {
			results = append(results, task)
		}
	}
	return results
}

// FilterTasksByDueDateRange filters tasks by due date range
func (tm *TaskManager) FilterTasksByDueDateRange(startDate, endDate time.Time) []*Task {
	tm.mu.RLock()
	defer tm.mu.RUnlock()

	var results []*Task
	for _, task := range tm.tasks {
		dueDate := task.GetDueDate()
		if (dueDate.Equal(startDate) || dueDate.After(startDate)) &&
			(dueDate.Equal(endDate) || dueDate.Before(endDate)) {
			results = append(results, task)
		}
	}
	return results
}

// MarkTaskAsCompleted marks a task as completed and moves it to history
func (tm *TaskManager) MarkTaskAsCompleted(taskId string) bool {
	tm.mu.Lock()
	defer tm.mu.Unlock()

	task, exists := tm.tasks[taskId]
	if !exists {
		return false
	}

	task.SetStatus(TaskStatusCompleted)

	// Add to user's task history
	if task.GetAssignedTo() != nil {
		userId := task.GetAssignedTo().GetId()
		tm.taskHistory[userId] = append(tm.taskHistory[userId], task)
	}

	return true
}

// GetTaskHistory retrieves task history for a user
func (tm *TaskManager) GetTaskHistory(userId string) []*Task {
	tm.mu.RLock()
	defer tm.mu.RUnlock()

	history := tm.taskHistory[userId]
	if history == nil {
		return []*Task{}
	}

	// Return a copy to avoid external modification
	result := make([]*Task, len(history))
	copy(result, history)
	return result
}

// AddUser adds a user to the system
func (tm *TaskManager) AddUser(user *User) {
	tm.mu.Lock()
	defer tm.mu.Unlock()
	tm.users[user.GetId()] = user
}

// GetUser retrieves a user by ID
func (tm *TaskManager) GetUser(userId string) *User {
	tm.mu.RLock()
	defer tm.mu.RUnlock()
	return tm.users[userId]
}

// AssignTask assigns a task to a user
func (tm *TaskManager) AssignTask(taskId, userId string) bool {
	tm.mu.Lock()
	defer tm.mu.Unlock()

	task, taskExists := tm.tasks[taskId]
	user, userExists := tm.users[userId]

	if !taskExists || !userExists {
		return false
	}

	// Remove from previous user if assigned
	if task.GetAssignedTo() != nil {
		task.GetAssignedTo().RemoveAssignedTask(task)
	}

	// Assign to new user
	user.AddAssignedTask(task)
	return true
}

// GetAllTasks returns all tasks
func (tm *TaskManager) GetAllTasks() []*Task {
	tm.mu.RLock()
	defer tm.mu.RUnlock()

	tasks := make([]*Task, 0, len(tm.tasks))
	for _, task := range tm.tasks {
		tasks = append(tasks, task)
	}
	return tasks
}

// Helper function to check if a string contains a substring (case-insensitive)
func contains(str, substr string) bool {
	return len(str) >= len(substr) && stringContains(str, substr)
}

func stringContains(s, substr string) bool {
	for i := 0; i <= len(s)-len(substr); i++ {
		if s[i:i+len(substr)] == substr {
			return true
		}
	}
	return false
}
